File: advent_of_code/2023/01/second.c
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 1000

const char WORDS[10][6] = {"zero", "one", "two",   "three", "four",
                           "five", "six", "seven", "eight", "nine"};

const char DIGITS[10][2] = {"0", "1", "2", "3", "4", "5", "6", "7", "8", "9"};

int main() {
  unsigned sum = 0;

  FILE *fp = fopen("input.txt", "r");

  char line[MAX];

  while (fgets(line, MAX, fp) != NULL) {
    int len = strlen(line);
    int first = 0;
    int last = 0;
    bool found_first = false;
    bool found_last = false;
    for (size_t i = 0; i < len; i++) {
      if (found_first) {
        break;
      }
      for (size_t j = 0; j < 10; j++) {
        if (strncmp(line + i, WORDS[j], strlen(WORDS[j])) == 0) {
          first = j;
          found_first = true;
          break;
        }
        if (strncmp(line + i, DIGITS[j], 1) == 0) {
          first = j;
          found_first = true;
          break;
        }
      }
    }
    for (size_t i = len - 1; i >= 0; i--) {
      if (found_last) {
        break;
      }
      for (size_t j = 0; j < 10; j++) {
        const size_t len = strlen(WORDS[j]);
        if (strncmp(line + i - len, WORDS[j], len) == 0) {
          last = j;
          found_last = true;
          break;
        }
        if (strncmp(line + i - 1, DIGITS[j], 1) == 0) {
          last = j;
          found_last = true;
          break;
        }
      }
    }
    sum += first * 10 + last;
  }

  printf("%d\n", sum);

  fclose(fp);

  return 0;
}

File: advent_of_code/2023/01/second.js
const fs = require('node:fs')

const input = fs.readFileSync('./input.txt', 'utf8')

const lines = input.split('\n')

var sum = 0

const WORDS = ['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
const DIGITS = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']

for (const line of lines) {
  let found_first = false;
  let found_last = false;

  for(let begin = 0; begin < line.length; begin++) {
    if (found_first) break
    for (let i = 0; i < 10; i++) {
      const word = WORDS[i]
      if (line.substring(begin, begin + word.length) === word) {
        sum += i * 10
        found_first = true
        break
      }
    }
    for (let i = 0; i < 10; i++) {
      const word = DIGITS[i]
      if (line.substring(begin, begin + word.length) === word) {
        sum += i * 10
        found_first = true
        break
      }
    }
  }

  for(let end = line.length; end > 0; end--) {
    if (found_last) break
    for (let i = 0; i < 10; i++) {
      const word = WORDS[i]
      if (line.substring(end - word.length, end) === word) {
        sum += i
        found_last = true
        break
      }
    }
    for (let i = 0; i < 10; i++) {
      const word = DIGITS[i]
      if (line.substring(end - word.length, end) === word) {
        sum += i
        found_last = true
        break
      }
    }
  }
}

console.log(sum)